stan/math/rev/core/operator_addition.hpp

```diff
@@ -3,38 +3,59 @@
 
 #include <stan/math/prim/meta.hpp>
 #include <stan/math/rev/core/var.hpp>
-#include <stan/math/rev/core/vv_vari.hpp>
-#include <stan/math/rev/core/vd_vari.hpp>
+#include <stan/math/rev/core/op_vari.hpp>
 #include <stan/math/prim/fun/constants.hpp>
 #include <stan/math/prim/fun/is_any_nan.hpp>
+#include <type_traits>
 
 namespace stan {
 namespace math {
 
 namespace internal {
-class add_vv_vari : public op_vv_vari {
+
+template <typename T>
+struct is_vari : bool_constant<std::is_base_of<
+                     vari, std::remove_pointer_t<std::decay_t<T>>>::value> {};
+
+template <typename T>
+using require_vari_t = require_t<is_vari<T>>;
+
+template <typename... Types>
+using require_all_vari_t = require_all_t<is_vari<Types>...>;
+
+template <typename T1, typename T2, typename = void>
+class add_vari {
+  static_assert(1, "If you see this please report a bug!");
+};
+
+template <typename T1, typename T2>
+class add_vari<T1, T2, require_all_vari_t<T1, T2>> : public op_vari<T1, T2> {
  public:
-  add_vv_vari(vari* avi, vari* bvi)
-      : op_vv_vari(avi->val_ + bvi->val_, avi, bvi) {}
+  add_vari(T1 avi, T2 bvi) : op_vari<T2, T1>(avi->val_ + bvi->val_, avi, bvi) {}
   void chain() {
-    if (unlikely(is_any_nan(avi_->val_, bvi_->val_))) {
-      avi_->adj_ = NOT_A_NUMBER;
-      bvi_->adj_ = NOT_A_NUMBER;
+    if (unlikely(is_any_nan(std::get<0>(this->vi())->val_,
+                            std::get<1>(this->vi())->val_))) {
+      std::get<0>(this->vi())->adj_ = NOT_A_NUMBER;
+      std::get<1>(this->vi())->adj_ = NOT_A_NUMBER;
     } else {
-      avi_->adj_ += adj_;
-      bvi_->adj_ += adj_;
+      std::get<0>(this->vi())->adj_ += this->adj_;
+      std::get<1>(this->vi())->adj_ += this->adj_;
```

Comment:
This all seems much more complicated than the code was before.  What's the advantage of this new scheme?  As far as I can tell, efficiency will be the same.

Reply:
It's wayyy more complicated than before. There's no computational advantage here, the idea is that if we can get this to work for `matrix_cl<var>` types then we won't have to rewrite a lot of things and can push elementwise operations onto the gpu.

There's a real tradeoff here between duplication and complication. imo I'd rather have a new pattern that's more complicated but consistent over having to maintain duplicate code. The other way to do this would be to template vv_vari etc., but I'm not sure if I like that. `std::get<>` is certainly weird, but while the call site is weird the pattern itself is simple and removes a good few specialized classes

Reply:
This could be simplified with a pair of functions in `op_vari`:
```
template <std::size_t Ind>
auto& arg_val (){
    return std::get<Ind>(this->vi())->val_;
}


template <std::size_t Ind>
auto& arg_adj (){
    return std::get<Ind>(this->vi())->adj_;
}
```

Regardless it is hard to see the point of this PR. I know it is a lot to ask, but a prototype of how this would work with `matrix_cl<var>` would be super helpful.

Reply:
Or if you don't like writing `this->template arg_val<2>()` you can even hardcode options for lets say up to 3 arguments.

Reply:
> Regardless it is hard to see the point of this PR. I know it is a lot to ask, but a prototype of how this would work with matrix_cl<var> would be super helpful.

Yeah it's still very vague. My intent here was to try to write out the operator methods such that we could pass in a `matrix_cl<var>`. It's reasonable to want to see how we would use this before taking on any complication costs

What I'm thinking is that for 

```cpp
inline matrix_cl<var> operator+(const matrix_cl<var>& a, const matrix_cl<var>& b) {
  return {new internal::add_vari<matrix_cl<var>*, matrix_cl<var>*>(a, b)};
}
```

Say the stan compiler could generate the silly/trivial code below

```cpp
Eigen::Matrix<var, -1, -1> x = from_matrix_cl(to_matrix_cl(A) + to_matrix_cl(B))
```

1. `to_matrix_cl(A/B)` gives back a `matrix_cl<var>` that then calls the above `operator+`. 
2. `operator+` would add that `add_vari` to the returned `matrix_cl<var>` stored as `vari*`. When we do the reverse mode that `add_vari` exists on the autodiff stack. 
4. When `grad()` is called and starts executing `.chain()` methods on the autodiff stack...
5. (???)

For (5) my main Q is where those `matrix_cl<var>`s we point to stored so RAII doesn't clean up a matrix_cl<var> before we call chain

  - One thing we could do is store the `matrix_cl<var>`'s memory in the context (kind of abusing the context but we made it a singleton and it's there)
  - Is there something smarter we can do since we know that an `Eigen::Matrix<var,-1,-1>`'s data is stored on our stack? The `val_` part is `const` so we'll have to copy that, but we could take ownership of the `adj_`
  - And do we want to try to be lazy about data movement here so that the GPU doesn't crap out from lack of dram?

Reply:
> One thing we could do is store the matrix_cl<var>'s memory in the context (kind of abusing the context but we made it a singleton and it's there)

I think having a separate chain stack for `matrix_cl<var>`s would be preferable.

> The val_ part is const so we'll have to copy that, but we could take ownership of the adj_

I don't follow that part.

> And do we want to try to be lazy about data movement here so that the GPU doesn't crap out from lack of dram?

That is probably a good idea. It might help to wait until reverse sweep before adjoint matrix_cl is allocated.

Reply:
>> The val_ part is const so we'll have to copy that, but we could take ownership of the adj_

> I don't follow that part.

If we wanted to not have an OpenCL stack and use stan's stack I think we could do some trick with [enqueueMapBuffer](https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clEnqueueMapBuffer.html) and enqueueUnMapBuffer. Though I'm not sure if that's a thing we can do

Reply:
We could do it, but I think OpenCL stack is the way to go. This mapping seems to be just a conveniance function that hides host<->device transfers.

Reply:
For GPUs it's just a wrapper effectively. But for cpus the map buffer can take ownership of the memory so we avoid a copy. I think using them would let us use the OpenCL stuff for smaller problems while still getting parallelization. Though what we really want is shared virtual memory. It looks like Nvidia has started supporting that (from windows [release notes](http://us.download.nvidia.com/Windows/442.19/442.19-win10-win8-win7-release-notes.pdf))

https://software.intel.com/en-us/articles/getting-the-most-from-opencl-12-how-to-increase-performance-by-minimizing-buffer-copies-on-intel-processor-graphics

Reply:
> I think using them would let us use the OpenCL stuff for smaller problems while still getting parallelization

Running one chain per core is far easier. I think it is also likely to be more efficient.

> Though what we really want is shared virtual memory

That just means same address space. Unless you are thinking about CPUs or integrated GPUs copying still needs to happen.